src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs

```diff
@@ -190,49 +205,165 @@ private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
-                }
-                else if (ancestor is IArgumentOperation argumentAncestor && argumentAncestor.Parent is IInvocationOperation argumentInvocationAncestor)
+                else if (context.Operation is IBinaryOperation binaryOperation)
                 {
-                    diagnosableInvocation = argumentInvocationAncestor;
-                }
-                else
-                {
-                    return;
+                    AnalyzeBinaryOperation(context, binaryOperation, stringType);
                 }
+            },
+                OperationKind.Invocation, // a.ToLower().Method()
+                OperationKind.Binary // a.ToLower() == b.ToLower()
+            );
+        }
 
-                IMethodSymbol diagnosableMethod = diagnosableInvocation.TargetMethod;
+        private static void AnalyzeInvocation(OperationAnalysisContext context, IInvocationOperation caseChangingInvocation, INamedTypeSymbol stringType,
+            IMethodSymbol containsStringMethod, IMethodSymbol startsWithStringMethod, IMethodSymbol compareToStringMethod,
+            IMethodSymbol indexOfStringMethod, IMethodSymbol indexOfStringInt32Method, IMethodSymbol indexOfStringInt32Int32Method)
+        {
+            if (!IsOffendingMethod(caseChangingInvocation, stringType, out string? offendingMethodName))
+            {
+                return;
+            }
 
-                if (diagnosableMethod.Equals(containsStringMethod) ||
-                    diagnosableMethod.Equals(startsWithStringMethod) ||
-                    diagnosableMethod.Equals(indexOfStringMethod) ||
-                    diagnosableMethod.Equals(indexOfStringInt32Method) ||
-                    diagnosableMethod.Equals(indexOfStringInt32Int32Method))
-                {
-                    context.ReportDiagnostic(diagnosableInvocation.CreateDiagnostic(RecommendCaseInsensitiveStringComparisonRule, diagnosableMethod.Name));
-                }
-                else if (diagnosableMethod.Equals(compareToStringMethod))
+            if (!TryGetInvocationWithoutParentheses(caseChangingInvocation, out IInvocationOperation? diagnosableInvocation))
+            {
+                return;
+            }
+
+            string caseChangingApproach = GetCaseChangingApproach(offendingMethodName);
+
+            ImmutableDictionary<string, string?> dict = new Dictionary<string, string?>()
+            {
+                { CaseChangingApproachName, caseChangingApproach }
+            }.ToImmutableDictionary();
+
+            IMethodSymbol diagnosableMethod = diagnosableInvocation.TargetMethod;
+
+            if (diagnosableMethod.Equals(containsStringMethod) ||
+                diagnosableMethod.Equals(startsWithStringMethod) ||
+                diagnosableMethod.Equals(indexOfStringMethod) ||
+                diagnosableMethod.Equals(indexOfStringInt32Method) ||
+                diagnosableMethod.Equals(indexOfStringInt32Int32Method))
```

Comment:
It seems `IMethodSymbol containsStringMethod, IMethodSymbol startsWithStringMethod, IMethodSymbol indexOfStringMethod, IMethodSymbol indexOfStringInt32Method, IMethodSymbol indexOfStringInt32Int32Method` all these IMethodSymbols created and used similarly, could we add them into a `ImmutableDictionary<IMethodSymbol>`  and check if it contains the diagnosableMethod?

Reply:
No, sorry I meant `ImmutableList` not `ImmutableDictionary`, I meant populate them once and add them into a `ImmutableList<IMethodSymbol> list` and instead of these multiple ORs check `if (list.Contains(diagnosableMethod))`

Reply:
I think we can save some space by not allocating a new ImmutableList.

Reply:
Even though it would be faster for lookup, and it will be cleaner to use without passing so many parameters. Though my concern was more about that all symbols are loaded one by one using Linq, they could have populated with one or 2 loops, even without adding them into a `ImmutableArray`. 
https://github.com/dotnet/roslyn-analyzers/blob/4e17f73a2eec06d82862a123f6fdbd52382adb8a/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/RecommendCaseInsensitiveStringComparison.Analyzer.cs#L83-L204
But as that logic was not added with this PR so its not have to be handled in this PR